[BUG] Regression: undefined reference to `dladdr'

Builds for several shards are failing with current nightly build (`Crystal 0.27.1-dev [530b91345] (2019-01-09)`). Looks like a codegen bug.

* https://circleci.com/gh/crystal-lang/shards/142
* https://circleci.com/gh/straight-shoota/crinja/250
* https://circleci.com/gh/straight-shoota/sass.cr/155

The error message with shards and crinja goes like this:
```
/usr/bin/crystal spec
_main.o: In function `*Exception+@Exception#inspect_with_backtrace<IO::FileDescriptor>:IO::FileDescriptor':
main_module:(.text+0x7a8e): undefined reference to `dladdr'
_main.o: In function `*CallStack::decode_frame<Pointer(Void), Pointer(Void)>:(Tuple(Int64, Pointer(UInt8)) | Nil)':
main_module:(.text+0x16c53): undefined reference to `dladdr'
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc "${@}" -o '/root/.cache/crystal/usr-share-crystal-src-ecr-process.cr/macro_run'  -rdynamic  -lpcre -lgc -lpthread /usr/share/crystal/src/ext/libcrystal.a -levent -lrt -L/usr/lib -L/usr/local/lib`
```

sass.cr has a different error, but it's most likely related, given that it pops up in the same nightly release:

```
crystal spec
/usr/bin/ld: C-allS-tack.o: undefined reference to symbol 'dladdr@@GLIBC_2.2.5'
//lib/x86_64-linux-gnu/libdl.so.2: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc "${@}" -o '/root/.cache/crystal/crystal-run-spec.tmp'  -rdynamic  -lsass -lpcre -lm -lgc -lpthread /usr/share/crystal/src/ext/libcrystal.a -levent -lrt -L/usr/lib -L/usr/local/lib`
Exited with code 1
```

The only PR merged since yesterday's nightly build is #7282

/cc @RX14

I can confirm that the issue occurs after merging #7282.

Before PR:

```
root@627bdde4e0eb:/src# crystal --version
Crystal 0.27.1-dev [03290575b] (2019-01-16)

LLVM: 4.0.0
Default target: x86_64-unknown-linux-gnu
root@627bdde4e0eb:/src# make clean crystal
Using /usr/bin/llvm-config-4.0 [version=4.0.0]
rm -rf .build
rm -rf ./docs
rm -rf src/llvm/ext/llvm_ext.o
rm -rf src/ext/sigfault.o src/ext/libcrystal.a
g++ -c  -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc -I/usr/lib/llvm-4.0/include -std=c++0x -gsplit-dwarf -Wl,-fuse-ld=gold -fPIC -fvisibility-inlines-hidden -Wall -W -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wno-maybe-uninitialized -Wdelete-non-virtual-dtor -Wno-comment -Werror=date-time -std=c++11 -ffunction-sections -fdata-sections -O2 -g -DNDEBUG  -fno-exceptions -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
cc -fPIC    -c -o src/ext/sigfault.o src/ext/sigfault.c
ar -rcs src/ext/libcrystal.a src/ext/sigfault.o
CRYSTAL_CONFIG_PATH="/src/src" CRYSTAL_CONFIG_BUILD_COMMIT="076d087" ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib
```

After PR

```
root@627bdde4e0eb:/src# crystal --version
Crystal 0.27.1-dev [530b91345] (2019-01-16)

LLVM: 4.0.0
Default target: x86_64-unknown-linux-gnu
root@627bdde4e0eb:/src# make clean crystal
Using /usr/bin/llvm-config-4.0 [version=4.0.0]
rm -rf .build
rm -rf ./docs
rm -rf src/llvm/ext/llvm_ext.o
rm -rf src/ext/sigfault.o src/ext/libcrystal.a
g++ -c  -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc -I/usr/lib/llvm-4.0/include -std=c++0x -gsplit-dwarf -Wl,-fuse-ld=gold -fPIC -fvisibility-inlines-hidden -Wall -W -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wno-maybe-uninitialized -Wdelete-non-virtual-dtor -Wno-comment -Werror=date-time -std=c++11 -ffunction-sections -fdata-sections -O2 -g -DNDEBUG  -fno-exceptions -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
cc -fPIC    -c -o src/ext/sigfault.o src/ext/sigfault.c
ar -rcs src/ext/libcrystal.a src/ext/sigfault.o
CRYSTAL_CONFIG_PATH="/src/src" CRYSTAL_CONFIG_BUILD_COMMIT="076d087" ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib
_main.o: In function `*Exception+@Exception#inspect_with_backtrace<IO::FileDescriptor>:IO::FileDescriptor':
main_module:(.text+0x7a8e): undefined reference to `dladdr'
_main.o: In function `*CallStack::decode_frame<Pointer(Void), Pointer(Void)>:(Tuple(Int64, Pointer(UInt8)) | Nil)':
main_module:(.text+0x16cd3): undefined reference to `dladdr'
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc "${@}" -o '/root/.cache/crystal/src-src-ecr-process.cr/macro_run'  -rdynamic  -lpcre -lgc -lpthread /src/src/ext/libcrystal.a -levent -lrt -L/usr/lib -L/usr/local/lib`
Makefile:122: recipe for target '.build/crystal' failed
make: *** [.build/crystal] Error 1
```

After the nightly build the generated compiler is not checked. That is done in test-ecosystem which is not automated to run.

Besides a need for a fix on this, maybe the nightly builds should run a checkout and build the compiler as a smoke test. That would have catch this without the need of external projects.

@RX14 Maybe the context for running macros has changed after the PR and is missing `[LD_]LIBRARY_PATH`?

--- a/src/compiler/crystal/codegen/target.cr
+++ b/src/compiler/crystal/codegen/target.cr
@@ -8,7 +8,7 @@ class Crystal::Codegen::Target
   getter vendor : String
   getter environment : String
 
-  def initialize(target_triple = LLVM.default_target_triple)
+  def initialize(target_triple = Crystal::Config.default_target_triple)
     # Let LLVM convert the user-inputted target triple into at least a target
     # triple with the architecture, vendor and OS in the correct place.
     target_triple = LLVM.normalize_triple(target_triple.downcase)

